% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/PRSDS.R
\name{PRSDS}
\alias{PRSDS}
\title{Get Ploygenic Risk Score}
\usage{
PRSDS(resources, snp_threshold, snp_assoc, pgs_id, ...)
}
\arguments{
\item{resources}{\code{list} of all the VCF resources with biallelic genotype information. It is advised to 
have one VCF resource per chromosome, a big VCF file with all the information is always slower 
to use.}

\item{snp_threshold}{\code{numeric} (default \code{80}) Threshold to drop individuals. See details for 
further information.}

\item{...}{Corresponds to the prs_table table passed from the client. It is assembled if not NULL.}
}
\value{
\code{data.frame} were the rownames are the individuals. The columns found are: \cr
prs: Polygenic risk score per individual \cr
prs_nw: Polygenic risk score without weights (weight 1 for each risk allele) \cr
n_snps: Number of SNPs with information for each individual
}
\description{
Get Ploygenic Risk Score
}
\details{
This function resolves a list of resources subsetting them by the 
SNPs of risk, this does not ensure that all the SNPs of risk will be found on the 
data. From all the found SNPs of risk, if an individual has less than 'snp_threshold' (percetage)
of SNPs with data, it will be dropped (SNP with no data is marked on the VCF as ./.). If an individual 
passes this threshold filter but still has SNPs with no data, those SNPs will be counted on the 
polygenic risk score as non-risk-alleles, to take this infomation into account, the number of SNPs 
with data for each individual is returned as 'n_snps'.
}
