<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ dsOmics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dsOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/geno_exploration.html">OmicSHIELD bookdown</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addPhenoDataDS.html">addPhenoDataDS()</a></code> </p>
        </td>
        <td><p>Add Phenotype data to ExpressionSet</p></td>
      </tr><tr><td>
          <p><code><a href="alleleFrequencyDS.html">alleleFrequencyDS()</a></code> </p>
        </td>
        <td><p>Allelic frequency</p></td>
      </tr><tr><td>
          <p><code><a href="BAM2VCF.html">BAM2VCF()</a></code> </p>
        </td>
        <td><p>Variant discovery on a BAM file</p></td>
      </tr><tr><td>
          <p><code><a href="cellCountsDS.html">cellCountsDS()</a></code> </p>
        </td>
        <td><p>Estimate cell counts from methylation data</p></td>
      </tr><tr><td>
          <p><code><a href="createRSEDS.html">createRSEDS()</a></code> </p>
        </td>
        <td><p>create a RangedSummarizedExperiment from a RNAseq count table and 
a phenotypes table</p></td>
      </tr><tr><td>
          <p><code><a href="DESeq2DS.html">DESeq2DS()</a></code> </p>
        </td>
        <td><p>Differential expression analysis on the server-side</p></td>
      </tr><tr><td>
          <p><code><a href="dlaplace.html">dlaplace()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="dot-fittedValues_exponential.html">.fittedValues_exponential()</a></code> </p>
        </td>
        <td><p>Fit values for binomial model</p></td>
      </tr><tr><td>
          <p><code><a href="dot-fittedValues_linear.html">.fittedValues_linear()</a></code> </p>
        </td>
        <td><p>Fit values for linear model</p></td>
      </tr><tr><td>
          <p><code><a href="dot-retrievePGS.html">.retrievePGS()</a></code> </p>
        </td>
        <td><p>Internal function: Get PGS catalog table of polygenic risks</p></td>
      </tr><tr><td>
          <p><code><a href="edgeRDS.html">edgeRDS()</a></code> </p>
        </td>
        <td><p>Differential expression analysis on the server-side</p></td>
      </tr><tr><td>
          <p><code><a href="EGAhtsgetResourceClient.html">EGAhtsgetResourceClient</a></code> </p>
        </td>
        <td><p>EGA htsget resource client</p></td>
      </tr><tr><td>
          <p><code><a href="EGAhtsgetResourceResolver.html">EGAhtsgetResourceResolver</a></code> </p>
        </td>
        <td><p>EGA htsget resource resolver</p></td>
      </tr><tr><td>
          <p><code><a href="exactHWEDS.html">exactHWEDS()</a></code> </p>
        </td>
        <td><p>Hardy-Weinberg Equilibrium testing</p></td>
      </tr><tr><td>
          <p><code><a href="extractPhenoFromGDSDS.html">extractPhenoFromGDSDS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="fastAlleleFrequencyDS.html">fastAlleleFrequencyDS()</a></code> </p>
        </td>
        <td><p>Allelic frequency</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_ColSums.html">fastGWAS_ColSums()</a></code> </p>
        </td>
        <td><p>Get colsums needed for fast GWAS</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_getFitted.values.html">fastGWAS_getFitted.values()</a></code> </p>
        </td>
        <td><p>Obtain fitted values given coefficients, family and values table</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_getResiduals.html">fastGWAS_getResiduals()</a></code> </p>
        </td>
        <td><p>Obtain residual values given fitted values, family and objective variable</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_PHENO_removeNAindiv.html">fastGWAS_PHENO_removeNAindiv()</a></code> </p>
        </td>
        <td><p>Remove individuals from the phenotype that have NAs on any of the objective variable or covars</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_S.html">fastGWAS_S()</a></code> </p>
        </td>
        <td><p>Extract Geno iterator and Pheno information from GenotypeData GDS containers</p></td>
      </tr><tr><td>
          <p><code><a href="fastGWAS_S_means.html">fastGWAS_S_means()</a></code> </p>
        </td>
        <td><p>Get mean of the genotype by individual</p></td>
      </tr><tr><td>
          <p><code><a href="featureDataDS.html">featureDataDS()</a></code> </p>
        </td>
        <td><p>Retrieve information on features recorded in eSet-derived objects.</p></td>
      </tr><tr><td>
          <p><code><a href="featureNamesDS.html">featureNamesDS()</a></code> </p>
        </td>
        <td><p>Retrieve feature names from eSets.</p></td>
      </tr><tr><td>
          <p><code><a href="fvarLabelsDS.html">fvarLabelsDS()</a></code> </p>
        </td>
        <td><p>Retrieve names of annotation from eSets.</p></td>
      </tr><tr><td>
          <p><code><a href="GA4GHResourceClient.html">GA4GHResourceClient</a></code> </p>
        </td>
        <td><p>GA4GH htsget server resource client</p></td>
      </tr><tr><td>
          <p><code><a href="GA4GHResourceResolver.html">GA4GHResourceResolver</a></code> </p>
        </td>
        <td><p>GA4GH htsget server resource resolver</p></td>
      </tr><tr><td>
          <p><code><a href="GDSFileResourceClient.html">GDSFileResourceClient</a></code> </p>
        </td>
        <td><p>GDS file resource client</p></td>
      </tr><tr><td>
          <p><code><a href="GDSFileResourceResolver.html">GDSFileResourceResolver</a></code> </p>
        </td>
        <td><p>GDS file resource resolver</p></td>
      </tr><tr><td>
          <p><code><a href="gdsSubset2.html">gdsSubset2()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="genoDimensionsDS.html">genoDimensionsDS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="GenotypeDataDS.html">GenotypeDataDS()</a></code> </p>
        </td>
        <td><p>Class GenotypeData.</p></td>
      </tr><tr><td>
          <p><code><a href="geno_pca_pooled_addPC2GenoDS.html">geno_pca_pooled_addPC2GenoDS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="geno_pca_pooled_addPCDS.html">geno_pca_pooled_addPCDS()</a></code> </p>
        </td>
        <td><p>Add PCA results to geno</p></td>
      </tr><tr><td>
          <p><code><a href="getChromosomeNamesDS.html">getChromosomeNamesDS()</a></code> </p>
        </td>
        <td><p>Get names of chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="getSNPSbyGenDS.html">getSNPSbyGenDS()</a></code> </p>
        </td>
        <td><p>Subset GDS with gene(s)</p></td>
      </tr><tr><td>
          <p><code><a href="getVariable.html">getVariable()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="GWASDS.html">GWASDS()</a></code> </p>
        </td>
        <td><p>Genome-wide association analysis (GWAS) on the server-side</p></td>
      </tr><tr><td>
          <p><code><a href="Laplace_noise_generator.html">Laplace_noise_generator()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="limmaDS.html">limmaDS()</a></code> </p>
        </td>
        <td><p>Differential expression analysis using limma on the server-side</p></td>
      </tr><tr><td>
          <p><code><a href="make_valid_namesDS.html">make_valid_namesDS()</a></code> </p>
        </td>
        <td><p>Rename column names and character variables</p></td>
      </tr><tr><td>
          <p><code><a href="nFeaturesDS.html">nFeaturesDS()</a></code> </p>
        </td>
        <td><p>Retrieve number of features from eSets and RangedSummarizedExperiments.</p></td>
      </tr><tr><td>
          <p><code><a href="nSamplesDS.html">nSamplesDS()</a></code> </p>
        </td>
        <td><p>Retrieve number of samples from eSets and RangedSummarizedExperiments.</p></td>
      </tr><tr><td>
          <p><code><a href="Opal2FileResourceGetter.html">Opal2FileResourceGetter</a></code> </p>
        </td>
        <td><p>Opal file resource getter</p></td>
      </tr><tr><td>
          <p><code><a href="PCADS.html">PCADS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="PCASNPSDS.html">PCASNPSDS()</a></code> </p>
        </td>
        <td><p>Principal Component Analysis (PCA) on SNP genotype data</p></td>
      </tr><tr><td>
          <p><code><a href="plinkDS.html">plinkDS()</a></code> </p>
        </td>
        <td><p>Genome-wide association analysis (GWAS) using PLINK</p></td>
      </tr><tr><td>
          <p><code><a href="plotPCADS.html">plotPCADS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plotPCASNPSDS.html">plotPCASNPSDS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="PRSDS.html">PRSDS()</a></code> </p>
        </td>
        <td><p>Get Ploygenic Risk Score</p></td>
      </tr><tr><td>
          <p><code><a href="PRSDS_aux.html">PRSDS_aux()</a></code> </p>
        </td>
        <td><p>Auxiliary function to merge the PRS results to a table by ID</p></td>
      </tr><tr><td>
          <p><code><a href="removeOutliersDS.html">removeOutliersDS()</a></code> </p>
        </td>
        <td><p>Filter potential CpG outliers</p></td>
      </tr><tr><td>
          <p><code><a href="rlaplace.html">rlaplace()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="RNAseqPreprocDS.html">RNAseqPreprocDS()</a></code> </p>
        </td>
        <td><p>Filter Genes By Expression Level</p></td>
      </tr><tr><td>
          <p><code><a href="selFeatureDS.html">selFeatureDS()</a></code> </p>
        </td>
        <td><p>Subset the information of a given feature in a eSet object.</p></td>
      </tr><tr><td>
          <p><code><a href="selSNPDS.html">selSNPDS()</a></code> </p>
        </td>
        <td><p>Subset the information of a given SNP in a GDS object.</p></td>
      </tr><tr><td>
          <p><code><a href="snptestDS.html">snptestDS()</a></code> </p>
        </td>
        <td><p>Interface to run SNPtest commands on a ssh connection (with SNPtest 2.5.2 installed)</p></td>
      </tr><tr><td>
          <p><code><a href="subsetExpressionSetDS.html">subsetExpressionSetDS()</a></code> </p>
        </td>
        <td><p>Subset ExpressionSet</p></td>
      </tr><tr><td>
          <p><code><a href="subsetGenoDS.html">subsetGenoDS()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="svdPartial.html">svdPartial()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="varLabelsDS.html">varLabelsDS()</a></code> </p>
        </td>
        <td><p>Retrieve feature names from eSets.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yannick Marcon, Xavier Esriba-Montagut, Juan R. Gonzalez.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

